{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1. Explain why we have to use the Exception class while creating a Custom Exception. \n",
    "Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# When creating a custom exception in Python, it is best practice to inherit from the built-in Exception class. This is because the Exception class serves as the base class for all built-in exceptions in Python. By inheriting from this class, we can take advantage of all the functionality provided by the Exception class, such as error messages and tracebacks.\n",
    "\n",
    "# Inheriting from the Exception class also ensures that our custom exception can be caught by the same try/except blocks that catch other built-in exceptions. This is because these blocks typically catch all exceptions that inherit from the Exception class.\n",
    "\n",
    "# Furthermore, by inheriting from the Exception class, we can customize the behavior of our custom exception by overriding methods such as str() and repr(). This allows us to provide more informative error messages and control how the exception is displayed to the user.\n",
    "\n",
    "# In summary, using the Exception class as the base class for a custom exception in Python provides a number of benefits, including compatibility with existing try/except blocks, access to built-in exception functionality, and the ability to customize the behavior of the exception."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BaseException\n",
      " BaseExceptionGroup\n",
      "  ExceptionGroup\n",
      " Exception\n",
      "  ArithmeticError\n",
      "   FloatingPointError\n",
      "   OverflowError\n",
      "   ZeroDivisionError\n",
      "    DivisionByZero\n",
      "    DivisionUndefined\n",
      "   DecimalException\n",
      "    Clamped\n",
      "    Rounded\n",
      "     Underflow\n",
      "     Overflow\n",
      "    Inexact\n",
      "     Underflow\n",
      "     Overflow\n",
      "    Subnormal\n",
      "     Underflow\n",
      "    DivisionByZero\n",
      "    FloatOperation\n",
      "    InvalidOperation\n",
      "     ConversionSyntax\n",
      "     DivisionImpossible\n",
      "     DivisionUndefined\n",
      "     InvalidContext\n",
      "  AssertionError\n",
      "  AttributeError\n",
      "   FrozenInstanceError\n",
      "  BufferError\n",
      "  EOFError\n",
      "   IncompleteReadError\n",
      "  ImportError\n",
      "   ModuleNotFoundError\n",
      "    PackageNotFoundError\n",
      "   ZipImportError\n",
      "  LookupError\n",
      "   IndexError\n",
      "   KeyError\n",
      "    NoSuchKernel\n",
      "    UnknownBackend\n",
      "   CodecRegistryError\n",
      "  MemoryError\n",
      "  NameError\n",
      "   UnboundLocalError\n",
      "  OSError\n",
      "   BlockingIOError\n",
      "   ChildProcessError\n",
      "   ConnectionError\n",
      "    BrokenPipeError\n",
      "    ConnectionAbortedError\n",
      "    ConnectionRefusedError\n",
      "    ConnectionResetError\n",
      "     RemoteDisconnected\n",
      "   FileExistsError\n",
      "   FileNotFoundError\n",
      "   InterruptedError\n",
      "    InterruptedSystemCall\n",
      "   IsADirectoryError\n",
      "   NotADirectoryError\n",
      "   PermissionError\n",
      "   ProcessLookupError\n",
      "   TimeoutError\n",
      "   UnsupportedOperation\n",
      "   herror\n",
      "   gaierror\n",
      "   SSLError\n",
      "    SSLCertVerificationError\n",
      "    SSLZeroReturnError\n",
      "    SSLWantWriteError\n",
      "    SSLWantReadError\n",
      "    SSLSyscallError\n",
      "    SSLEOFError\n",
      "   Error\n",
      "    SameFileError\n",
      "   SpecialFileError\n",
      "   ExecError\n",
      "   ReadError\n",
      "   URLError\n",
      "    HTTPError\n",
      "    ContentTooShortError\n",
      "   BadGzipFile\n",
      "  ReferenceError\n",
      "  RuntimeError\n",
      "   NotImplementedError\n",
      "    ZMQVersionError\n",
      "    StdinNotImplementedError\n",
      "   RecursionError\n",
      "   _DeadlockError\n",
      "   BrokenBarrierError\n",
      "   BrokenExecutor\n",
      "    BrokenThreadPool\n",
      "   SendfileNotAvailableError\n",
      "   BrokenBarrierError\n",
      "   ExtractionError\n",
      "   VariableError\n",
      "  StopAsyncIteration\n",
      "  StopIteration\n",
      "  SyntaxError\n",
      "   IndentationError\n",
      "    TabError\n",
      "  SystemError\n",
      "   CodecRegistryError\n",
      "  TypeError\n",
      "   FloatOperation\n",
      "   MultipartConversionError\n",
      "  ValueError\n",
      "   UnicodeError\n",
      "    UnicodeDecodeError\n",
      "    UnicodeEncodeError\n",
      "    UnicodeTranslateError\n",
      "   UnsupportedOperation\n",
      "   JSONDecodeError\n",
      "   SSLCertVerificationError\n",
      "   Error\n",
      "   UnsupportedDigestmodError\n",
      "   IllegalMonthError\n",
      "   IllegalWeekdayError\n",
      "   ParserError\n",
      "   MessageDefect\n",
      "    NoBoundaryInMultipartDefect\n",
      "    StartBoundaryNotFoundDefect\n",
      "    CloseBoundaryNotFoundDefect\n",
      "    FirstHeaderLineIsContinuationDefect\n",
      "    MisplacedEnvelopeHeaderDefect\n",
      "    MissingHeaderBodySeparatorDefect\n",
      "    MultipartInvariantViolationDefect\n",
      "    InvalidMultipartContentTransferEncodingDefect\n",
      "    UndecodableBytesDefect\n",
      "    InvalidBase64PaddingDefect\n",
      "    InvalidBase64CharactersDefect\n",
      "    InvalidBase64LengthDefect\n",
      "    HeaderDefect\n",
      "     InvalidHeaderDefect\n",
      "     HeaderMissingRequiredValue\n",
      "     NonPrintableDefect\n",
      "     ObsoleteHeaderDefect\n",
      "     NonASCIILocalPartDefect\n",
      "     InvalidDateDefect\n",
      "   ClassNotFound\n",
      "   ClipboardEmpty\n",
      "   MacroToEdit\n",
      "   InvalidFileException\n",
      "   UnequalIterablesError\n",
      "   InvalidVersion\n",
      "   _InvalidELFFileHeader\n",
      "   InvalidWheelFilename\n",
      "   InvalidSdistFilename\n",
      "   InvalidSpecifier\n",
      "   InvalidMarker\n",
      "   UndefinedComparison\n",
      "   UndefinedEnvironmentName\n",
      "   InvalidRequirement\n",
      "    RequirementParseError\n",
      "   InvalidVersion\n",
      "  Warning\n",
      "   BytesWarning\n",
      "   DeprecationWarning\n",
      "    ProvisionalWarning\n",
      "   EncodingWarning\n",
      "   FutureWarning\n",
      "    ProvisionalCompleterWarning\n",
      "   ImportWarning\n",
      "   PendingDeprecationWarning\n",
      "   ResourceWarning\n",
      "   RuntimeWarning\n",
      "    ProactorSelectorThreadWarning\n",
      "    UnknownTimezoneWarning\n",
      "    PEP440Warning\n",
      "   SyntaxWarning\n",
      "   UnicodeWarning\n",
      "   UserWarning\n",
      "    GetPassWarning\n",
      "    FormatterWarning\n",
      "   DeprecatedTzFormatWarning\n",
      "   PkgResourcesDeprecationWarning\n",
      "  ExceptionGroup\n",
      "  _OptionError\n",
      "  _Error\n",
      "  error\n",
      "  Error\n",
      "  SubprocessError\n",
      "   CalledProcessError\n",
      "   TimeoutExpired\n",
      "  ZMQBaseError\n",
      "   ZMQError\n",
      "    ContextTerminated\n",
      "    Again\n",
      "    InterruptedSystemCall\n",
      "   ZMQBindError\n",
      "   NotDone\n",
      "  Error\n",
      "  error\n",
      "  PickleError\n",
      "   PicklingError\n",
      "   UnpicklingError\n",
      "  _Stop\n",
      "  TokenError\n",
      "  StopTokenizing\n",
      "  Error\n",
      "   CancelledError\n",
      "   InvalidStateError\n",
      "  _GiveupOnSendfile\n",
      "  Incomplete\n",
      "  ClassFoundException\n",
      "  EndOfBlock\n",
      "  InvalidStateError\n",
      "  LimitOverrunError\n",
      "  QueueEmpty\n",
      "  QueueFull\n",
      "  error\n",
      "  LZMAError\n",
      "  RegistryError\n",
      "  _GiveupOnFastCopy\n",
      "  TraitError\n",
      "  Empty\n",
      "  Full\n",
      "  ArgumentError\n",
      "  COMError\n",
      "  ReturnValueIgnoredError\n",
      "  ArgumentError\n",
      "  ArgumentTypeError\n",
      "  ConfigError\n",
      "   ConfigLoaderError\n",
      "    ArgumentError\n",
      "   ConfigFileNotFound\n",
      "  ConfigurableError\n",
      "   MultipleInstanceError\n",
      "  ApplicationError\n",
      "  InvalidPortNumber\n",
      "  NoIPAddresses\n",
      "  Error\n",
      "  BadZipFile\n",
      "  LargeZipFile\n",
      "  MessageError\n",
      "   MessageParseError\n",
      "    HeaderParseError\n",
      "    BoundaryError\n",
      "   MultipartConversionError\n",
      "   CharsetError\n",
      "  DuplicateKernelError\n",
      "  ErrorDuringImport\n",
      "  NotOneValueFound\n",
      "  KnownIssue\n",
      "  VerifierFailure\n",
      "  CannotEval\n",
      "  OptionError\n",
      "  BdbQuit\n",
      "  Restart\n",
      "  FindCmdError\n",
      "  HomeDirError\n",
      "  ProfileDirError\n",
      "  IPythonCoreError\n",
      "   TryNext\n",
      "   UsageError\n",
      "   StdinNotImplementedError\n",
      "  InputRejected\n",
      "  GetoptError\n",
      "  ErrorToken\n",
      "  PrefilterError\n",
      "  AliasError\n",
      "   InvalidAliasError\n",
      "  Error\n",
      "   InterfaceError\n",
      "   DatabaseError\n",
      "    InternalError\n",
      "    OperationalError\n",
      "    ProgrammingError\n",
      "    IntegrityError\n",
      "    DataError\n",
      "    NotSupportedError\n",
      "  Warning\n",
      "  SpaceInInput\n",
      "  DOMException\n",
      "   IndexSizeErr\n",
      "   DomstringSizeErr\n",
      "   HierarchyRequestErr\n",
      "   WrongDocumentErr\n",
      "   InvalidCharacterErr\n",
      "   NoDataAllowedErr\n",
      "   NoModificationAllowedErr\n",
      "   NotFoundErr\n",
      "   NotSupportedErr\n",
      "   InuseAttributeErr\n",
      "   InvalidStateErr\n",
      "   SyntaxErr\n",
      "   InvalidModificationErr\n",
      "   NamespaceErr\n",
      "   InvalidAccessErr\n",
      "   ValidationErr\n",
      "  ValidationError\n",
      "  EditReadOnlyBuffer\n",
      "  _Retry\n",
      "  InvalidLayoutError\n",
      "  HeightIsUnknownError\n",
      "  GuardRejection\n",
      "  ParserSyntaxError\n",
      "  InternalParseError\n",
      "  _PositionUpdatingFinished\n",
      "  SimpleGetItemNotFound\n",
      "  UncaughtAttributeError\n",
      "  HasNoContext\n",
      "  ParamIssue\n",
      "  _JediError\n",
      "   InternalError\n",
      "   WrongVersion\n",
      "   RefactoringError\n",
      "  OnErrorLeaf\n",
      "  InvalidPythonEnvironment\n",
      "  HTTPException\n",
      "   NotConnected\n",
      "   InvalidURL\n",
      "   UnknownProtocol\n",
      "   UnknownTransferEncoding\n",
      "   UnimplementedFileMode\n",
      "   IncompleteRead\n",
      "   ImproperConnectionState\n",
      "    CannotSendRequest\n",
      "    CannotSendHeader\n",
      "    ResponseNotReady\n",
      "   BadStatusLine\n",
      "    RemoteDisconnected\n",
      "   LineTooLong\n",
      "  InteractivelyDefined\n",
      "  KillEmbedded\n",
      "  Error\n",
      "   NoSuchProcess\n",
      "    ZombieProcess\n",
      "   AccessDenied\n",
      "   TimeoutExpired\n",
      "  Error\n",
      "  TimeoutExpired\n",
      "  TimeoutAbandoned\n",
      "  KeyReuseError\n",
      "  UnknownKeyError\n",
      "  LeakedCallbackError\n",
      "  BadYieldError\n",
      "  ReturnValueIgnoredError\n",
      "  Return\n",
      "  QueueEmpty\n",
      "  QueueFull\n",
      "  DebuggerInitializationError\n",
      "  ExpatError\n",
      "  Error\n",
      "   ProtocolError\n",
      "   ResponseError\n",
      "   Fault\n",
      "  error\n",
      "  com_error\n",
      "  internal_error\n",
      "  ParseBaseException\n",
      "   ParseException\n",
      "   ParseFatalException\n",
      "    ParseSyntaxException\n",
      "  RecursiveGrammarException\n",
      "  ResolutionError\n",
      "   VersionConflict\n",
      "    ContextualVersionConflict\n",
      "   DistributionNotFound\n",
      "   UnknownExtra\n",
      "  _Error\n",
      "  UnableToResolveVariableException\n",
      "  InvalidTypeInArgsException\n",
      " GeneratorExit\n",
      " KeyboardInterrupt\n",
      " SystemExit\n",
      " CancelledError\n",
      " AbortThread\n"
     ]
    }
   ],
   "source": [
    "def exe_hierarachy(exception_class,level=0):\n",
    "    print(\" \"*level+exception_class.__name__)\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        exe_hierarachy(subclass,level+1)\n",
    "\n",
    "exe_hierarachy(BaseException)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3. What errors are defined in the ArithmeticError class? Explain any two with an example. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[14], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# ZeroDivisionError: This error is raised when a number is divided by zero. For example, dividing a number by zero in Python results in an error\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m \u001b[39m5\u001b[39;49m\u001b[39m/\u001b[39;49m\u001b[39m0\u001b[39;49m\n",
      "\u001b[1;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "# ZeroDivisionError: This error is raised when a number is divided by zero. For example, dividing a number by zero in Python results in an error\n",
    "\n",
    "5/0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "math domain error",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[15], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# ArithmeticError: This is a catch-all exception for arithmetic-related errors that don't have a more specific exception defined. For example:\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mmath\u001b[39;00m\n\u001b[1;32m----> 4\u001b[0m math\u001b[39m.\u001b[39;49msqrt(\u001b[39m-\u001b[39;49m\u001b[39m1\u001b[39;49m)\n",
      "\u001b[1;31mValueError\u001b[0m: math domain error"
     ]
    }
   ],
   "source": [
    "# ArithmeticError: This is a catch-all exception for arithmetic-related errors that don't have a more specific exception defined. For example:\n",
    "\n",
    "import math\n",
    "math.sqrt(-1)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q4. Why LookupError class is used? Explain with an example KeyError and IndexError. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'d'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[16], line 11\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# The LookupError class is a built-in class in Python that serves as the base class for all exceptions that occur when a key or index is not found in a mapping or sequence.\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \n\u001b[0;32m      3\u001b[0m \u001b[39m# The main reason to use the LookupError class is to catch these types of exceptions in a generic way, without having to specify the exact type of exception that might occur. This allows us to write more flexible and reusable code that can handle a variety of lookup-related errors.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m      6\u001b[0m \n\u001b[0;32m      7\u001b[0m \u001b[39m# KeyError is raised when a key is not found in a dictionary or other mapping\u001b[39;00m\n\u001b[0;32m     10\u001b[0m my_dict \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39m\u001b[39ma\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m1\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mb\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m2\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mc\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m3\u001b[39m}\n\u001b[1;32m---> 11\u001b[0m my_dict[\u001b[39m'\u001b[39;49m\u001b[39md\u001b[39;49m\u001b[39m'\u001b[39;49m]\n",
      "\u001b[1;31mKeyError\u001b[0m: 'd'"
     ]
    }
   ],
   "source": [
    "# The LookupError class is a built-in class in Python that serves as the base class for all exceptions that occur when a key or index is not found in a mapping or sequence.\n",
    "\n",
    "# The main reason to use the LookupError class is to catch these types of exceptions in a generic way, without having to specify the exact type of exception that might occur. This allows us to write more flexible and reusable code that can handle a variety of lookup-related errors.\n",
    "\n",
    "# Two common exceptions that are derived from LookupError are KeyError and IndexError.\n",
    "\n",
    "# KeyError is raised when a key is not found in a dictionary or other mapping\n",
    "\n",
    "\n",
    "my_dict = {'a': 1, 'b': 2, 'c': 3}\n",
    "my_dict['d']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[17], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m my_list \u001b[39m=\u001b[39m [\u001b[39m1\u001b[39m, \u001b[39m2\u001b[39m, \u001b[39m3\u001b[39m]\n\u001b[1;32m----> 2\u001b[0m my_list[\u001b[39m3\u001b[39;49m]\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3]\n",
    "my_list[3]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q5. Explain ImportError. What is ModuleNotFoundError? \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In Python, ImportError is a built-in exception that is raised when a module or package cannot be imported. This can happen for a variety of reasons, such as:\n",
    "\n",
    "# The module or package does not exist\n",
    "# The module or package is not installed\n",
    "# The module or package is not in the search path\n",
    "# There is a syntax error or other issue in the module or package"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Be specific: Catch specific exceptions whenever possible rather than using a generic exception such as Exception. This helps in identifying the root cause of the error more easily and prevents hiding other errors that might occur in the code.\n",
    "\n",
    "# Keep it simple: Don't put too much code inside a try -except block. Keep the code within the try block to a minimum to avoid catching unrelated exceptions.\n",
    "\n",
    "# Use finally block: Always use a finally block to clean up resources such as closing files, sockets, and database connections. The finally block is executed regardless of whether an exception is raised or not .\n",
    "\n",
    "# Use context managers: Use context managers such as the with statement to ensure that resources are properly managed and cleaned up even in the event of an exception.\n",
    "\n",
    "# Log errors: Always log the error message along with other relevant details such as the timestamp, module, function name, and line number. This helps in debugging the error later.\n",
    "\n",
    "# Reraise exceptions: In some cases, it might be necessary to reraise an exception after handling it. This can be done using the raise statement without any arguments, which re-raises the last exception that was caught.\n",
    "\n",
    "# Handle exceptions at the right level: Catch exceptions at the right level of abstraction. For example, if an exception occurs while opening a file, catch the exception in the function that opens the file, not in the higher-level code that calls the function.\n",
    "\n",
    "# Be careful with try -except -else: Use the else block with caution. It is only executed if no exception is raised in the try block. However, it can be confusing if it is not clear what code should be in the try block and what code should be in the else block.\n",
    "\n",
    "# Use built-in exceptions whenever possible: Use the built-in exceptions whenever possible instead of creating custom exceptions. This makes the code more understandable and easier to maintain.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
